File: maze/mazegen.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <fstream>
#include <cstdlib>
using namespace std;

class DSU
{
public:
    int *parent;
    int *rank;
    int n;

    DSU(){

    };

    void init(int n)
    {
        this->n = n;
        parent = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++)
        {
            parent[i] = i;
            rank[i] = 0;
        }
    }

    ~DSU()
    {
        delete[] parent;
        delete[] rank;
    }

    int find(int x)
    {
        if (parent[x] == x)
            return x;
        return parent[x] = find(parent[x]);
    }

    void unite(int x, int y)
    {
        int s1 = find(x);
        int s2 = find(y);
        if (s1 != s2)
        {
            if (rank[s1] < rank[s2])
                parent[s1] = s2;
            else if (rank[s1] > rank[s2])
                parent[s2] = s1;
            else
            {
                parent[s1] = s2;
                rank[s2]++;
            }
        }
    }
};

class Maze
{
public:
    int n;
    int m;
    DSU dsu;
    vector<pair<int, int>> edges;
    vector<pair<int, int>> maze;
    ofstream fout;
    ofstream fout2;

    int start_x = 0;
    int start_y = 0;
    int end_x, end_y;

    Maze(int n, int m)
    {
        this->n = n;
        this->m = m;
        dsu.init(n * m);
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (i + 1 < n)
                    edges.push_back({i * m + j, (i + 1) * m + j});
                if (j + 1 < m)
                    edges.push_back({i * m + j, i * m + j + 1});
            }
        }
        // shuffle the edges in a random order
                srand(time(NULL));
        for (int i = 0; i < edges.size(); i++)
        {
            int j = rand() % edges.size();
            swap(edges[i], edges[j]);
        }
        // initialize fout to maze.txt
        fout.open("maze.txt");
        fout2.open("mazehist.txt");
        // start_x = rand() % m;
        // start_y = rand() % n;
        end_x = rand() % m;
        end_y = rand() % n;
    }

    void generate()
    {

        for (auto edge : edges)
        {
            int u = edge.first;
            int v = edge.second;
            if (dsu.find(u) != dsu.find(v))
            {
                dsu.unite(u, v);
                maze.push_back({u, v});
                print(false);
            }
        }
    }

    void print( bool boo)
    {
        vector<vector<char>> grid(2 * n + 1, vector<char>(2 * m + 1, '#'));
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                grid[2 * i + 1][2 * j + 1] = ' ';
            }
        }
        for (auto edge : maze)
        {
            int u = edge.first;
            int v = edge.second;
            int x1 = u / m;
            int y1 = u % m;
            int x2 = v / m;
            int y2 = v % m;
            if (x1 == x2)
            {
                grid[2 * x1 + 1][2 * min(y1, y2) + 2] = ' ';
            }
            else
            {
                grid[2 * min(x1, x2) + 2][2 * y1 + 1] = ' ';
            }
        }

        // pick a random start and end point

        grid[2 * start_x + 1][2 * start_y + 1] = 'S';
        grid[2 * end_x + 1][2 * end_y + 1] = 'G';

        // now write the grid to maze.txt

        for (int i = 0; i < 2 * n + 1; i++)
        {
            for (int j = 0; j < 2 * m + 1; j++)
            {
                (boo ? fout : fout2) << grid[i][j];
            }
            (boo ? fout : fout2) << endl;
        }

    }
};

int main()
{
    Maze maze(10, 10);
    maze.generate();
    maze.print(true);
    return 0;
}

File: maze/mazegen.py
import random

class DSU:
    def __init__(self):
        self.parent = []
        self.rank = []
        self.n = 0

    def init(self, n):
        self.n = n
        self.parent = [i for i in range(n)]
        self.rank = [0 for i in range(n)]

    def find(self, x):
        if self.parent[x] == x:
            return x
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def unite(self, x, y):
        s1 = self.find(x)
        s2 = self.find(y)
        if s1 != s2:
            if self.rank[s1] < self.rank[s2]:
                self.parent[s1] = s2
            elif self.rank[s1] > self.rank[s2]:
                self.parent[s2] = s1
            else:
                self.parent[s1] = s2
                self.rank[s2] += 1

class Maze:
    def __init__(self, n, m):
        self.n = n
        self.m = m
        self.dsu = DSU()
        self.edges = []
        self.maze = []
        self.start_x = 0
        self.start_y = 0
        self.end_x = 0
        self.end_y = 0
        self.history = []
        self.dsu.init(n * m)
        for i in range(n):
            for j in range(m):
                if i + 1 < n:
                    self.edges.append((i * m + j, (i + 1) * m + j))
                if j + 1 < m:
                    self.edges.append((i * m + j, i * m + j + 1))

        random.seed()
        for i in range(len(self.edges)):
            j = random.randint(0, len(self.edges) - 1)
            self.edges[i], self.edges[j] = self.edges[j], self.edges[i]

        self.end_x = random.randint(0, m - 1)
        self.end_y = random.randint(0, n - 1)

    def __len__(self):
        return len(self.maze)
    def generate(self):
        for edge in self.edges:
            u, v = edge
            if self.dsu.find(u) != self.dsu.find(v):
                self.dsu.unite(u, v)
                self.maze.append((u, v))
                self.history.append(self.get())
        self.maze = self.history[-1]
        
    def __getitem__(self, idx):
        return self.maze[idx]

    def get(self):
        grid = [['#' for j in range(2 * self.m + 1)] for i in range(2 * self.n + 1)]
        for i in range(self.n):
            for j in range(self.m):
                grid[2 * i + 1][2 * j + 1] = ' '

        for edge in self.maze:
            u, v = edge
            x1 = u // self.m
            y1 = u % self.m
            x2 = v // self.m
            y2 = v % self.m
            if x1 == x2:
                grid[2 * x1 + 1][2 * min(y1, y2) + 2] = ' '
            else:
                grid[2 * min(x1, x2) + 2][2 * y1 + 1] = ' '
        return grid
    
if __name__ == "__main__":
    maze = Maze(4, 4)
    maze.generate()
    maze = maze.get()
    print("done")